src/librustc_ast/token.rs

```diff
@@ -785,6 +785,26 @@ impl Nonterminal {
             NtTT(tt) => tt.span(),
         }
     }
+
+    /// This nonterminal looks like some specific enums from
+    /// `proc-macro-hack` and `procedural-masquerade` crates.
+    /// We need to maintain some special pretty-printing behavior for them due to incorrect
+    /// asserts in old versions of those crates and their wide use in the ecosystem.
+    /// See issue #73345 for more details.
+    /// FIXME: Remove this eventually.
+    pub fn tokenization_compatibility_hack(&self) -> bool {
+        if let NtItem(item) = self {
+            let name = item.ident.name;
+            if name == sym::ProceduralMasqueradeDummyType || name == sym::ProcMacroHack {
```

Comment:
:sweat_smile:

Reply:
Can we add an allow by default lint here? That way we can make it warn by default in 6 months acting as a nudge so that we can outright remove this code in a year.

Reply:
We go through this code for both old and new versions of these crates, so we cannot discern between good and bad cases.
If we produced a warning, it would just be produced for all uses of `proc-macro-hack` and `procedural-masquerade`.